src/collision/narrow_phase.rs

```diff
@@ -162,139 +183,457 @@ pub enum NarrowPhaseSet {
+    /// will be copied over for the new contacts. Using these impulses as the initial guess is referred to
+    /// as *warm starting*, and it can help the contact solver resolve overlap and stabilize much faster.
+    #[allow(clippy::type_complexity, clippy::too_many_arguments)]
+    pub fn compute_contact_pair(
+        &self,
+        collider1: &ColliderQueryItem<C>,
+        collider2: &ColliderQueryItem<C>,
+        max_distance: Scalar,
+        match_contacts: bool,
+    ) -> Option<Contacts> {
+        let position1 = collider1.current_position();
+        let position2 = collider2.current_position();
+
+        // TODO: It'd be good to persist the manifolds and let Parry match contacts.
+        //       This isn't currently done because it requires using Parry's contact manifold type.
+        // Compute the contact manifolds using the effective speculative margin.
+        let mut manifolds = collider1.shape.contact_manifolds(
+            collider2.shape,
+            position1,
+            *collider1.rotation,
+            position2,
+            *collider2.rotation,
+            max_distance,
+        );
+
+        // Get the previous contacts if there are any.
+        let previous_contacts = self
+            .collisions
+            .get_internal()
+            .get(&(collider1.entity, collider2.entity))
+            .or(self
+                .collisions
+                .get_internal()
+                .get(&(collider2.entity, collider1.entity)));
+
+        let mut total_normal_impulse = 0.0;
+        let mut total_tangent_impulse = default();
+
+        // Match contacts and copy previous contact impulses for warm starting the solver.
+        // TODO: This condition is pretty arbitrary, mainly to skip dense trimeshes.
+        //       If we let Parry handle contact matching, this wouldn't be needed.
+        if manifolds.len() <= 4 && match_contacts {
+            if let Some(previous_contacts) = previous_contacts {
+                // TODO: Cache this?
+                let distance_threshold = 0.1 * self.length_unit.0;
+
+                for manifold in manifolds.iter_mut() {
+                    for previous_manifold in previous_contacts.manifolds.iter() {
+                        manifold.match_contacts(&previous_manifold.contacts, distance_threshold);
+
+                        // Add contact impulses to total impulses.
+                        for contact in manifold.contacts.iter() {
+                            total_normal_impulse += contact.normal_impulse;
+                            total_tangent_impulse += contact.tangent_impulse;
+                        }
+                    }
+                }
+            }
+        }
 
         let contacts = Contacts {
-            entity1,
-            entity2,
+            entity1: collider1.entity,
+            entity2: collider2.entity,
+            body_entity1: collider1.parent.map(|p| p.get()),
+            body_entity2: collider2.parent.map(|p| p.get()),
```

Comment:
```suggestion
            body_entity1: collider1.parent.map(Parent::get),
            body_entity2: collider2.parent.map(Parent::get),
```

Reply:
This would be `ColliderParent::get`, not sure if it's cleaner than just the closure

Reply:
Fair enough